state/logdb/buf_test.go

```diff
@@ -0,0 +1,161 @@
+package logdb_test
+
+import (
+	"errors"
+	"time"
+
+	"github.com/juju/names"
+	"github.com/juju/testing"
+	jc "github.com/juju/testing/checkers"
+	gc "gopkg.in/check.v1"
+
+	"github.com/juju/juju/state"
+	"github.com/juju/juju/state/logdb"
+	coretesting "github.com/juju/juju/testing"
+)
+
+var _ = gc.Suite(&BufferedLoggerSuite{})
+
+type BufferedLoggerSuite struct {
+	testing.IsolationSuite
+
+	mock  mockLogger
+	clock *testing.Clock
+}
+
+func (s *BufferedLoggerSuite) SetUpTest(c *gc.C) {
+	s.IsolationSuite.SetUpTest(c)
+	s.mock = mockLogger{}
+	s.clock = testing.NewClock(time.Time{})
+}
+
+func (s *BufferedLoggerSuite) TestLogFlushes(c *gc.C) {
+	const bufsz = 3
+	b := logdb.NewBufferedLogger(&s.mock, bufsz, time.Minute, s.clock)
+	in := []state.LogRecord{{
+		Entity:  names.NewMachineTag("0"),
+		Message: "foo",
+	}, {
+		Entity:  names.NewMachineTag("0"),
+		Message: "bar",
+	}, {
+		Entity:  names.NewMachineTag("0"),
+		Message: "baz",
+	}}
+
+	err := b.Log(in[:2])
+	c.Assert(err, jc.ErrorIsNil)
+	s.mock.CheckNoCalls(c)
+
+	err = b.Log(in[2:])
+	c.Assert(err, jc.ErrorIsNil)
+	s.mock.CheckCalls(c, []testing.StubCall{
+		{"Log", []interface{}{in}},
+	})
+}
+
+func (s *BufferedLoggerSuite) TestLogFlushesMultiple(c *gc.C) {
+	const bufsz = 1
+	b := logdb.NewBufferedLogger(&s.mock, bufsz, time.Minute, s.clock)
+	in := []state.LogRecord{{
+		Entity:  names.NewMachineTag("0"),
+		Message: "foo",
+	}, {
+		Entity:  names.NewMachineTag("0"),
+		Message: "bar",
+	}, {
+		Entity:  names.NewMachineTag("0"),
+		Message: "baz",
+	}}
+
+	err := b.Log(in)
+	c.Assert(err, jc.ErrorIsNil)
+	s.mock.CheckCalls(c, []testing.StubCall{
+		{"Log", []interface{}{in[:1]}},
+		{"Log", []interface{}{in[1:2]}},
+		{"Log", []interface{}{in[2:]}},
+	})
+}
+
+func (s *BufferedLoggerSuite) TestTimerFlushes(c *gc.C) {
+	const bufsz = 10
+	const flushInterval = time.Minute
+	s.mock.called = make(chan []state.LogRecord)
+
+	b := logdb.NewBufferedLogger(&s.mock, bufsz, flushInterval, s.clock)
+	in := []state.LogRecord{{
+		Entity:  names.NewMachineTag("0"),
+		Message: "foo",
+	}, {
+		Entity:  names.NewMachineTag("0"),
+		Message: "bar",
+	}}
+
+	err := b.Log(in[:1])
+	c.Assert(err, jc.ErrorIsNil)
+	s.mock.CheckNoCalls(c)
+
+	// Advance, but not far enough to trigger the flush.
+	s.clock.WaitAdvance(30*time.Second, coretesting.LongWait, 1)
+	s.mock.CheckNoCalls(c)
+
+	// Log again; the timer should not have been reset.
+	err = b.Log(in[1:])
+	s.mock.CheckNoCalls(c)
+
+	// Advance to to the flush interval.
+	s.clock.Advance(30 * time.Second)
+	select {
+	case records := <-s.mock.called:
+		c.Assert(records, jc.DeepEquals, in)
+		s.mock.CheckCalls(c, []testing.StubCall{
+			{"Log", []interface{}{in}},
+		})
+	case <-time.After(coretesting.LongWait):
+		c.Fatal("timed out waiting for logs to be flushed")
+	}
+}
```

Comment:
The other parts you are missing are that it *doesn't* trigger any future logs
and what time we will be flushing after future logging. (the set back to nil and it getting set again once we see another log)

Also, we are missing a test that if you Log([]1010) items, then we will flush the 1000 immediately, but wait the 1minute to flush the remaining 10, but we guarantee the 10 will get flushed eventually.

And potentially the "log 2 items, then N > buffer". Maybe that is just combined with the previous test. But the general "some stuff is queued, we overflow the buffer, but still have stuff queued".